core/types/typemaps.cc

```diff
@@ -56,168 +56,88 @@ TypePtr TypeVar::_approximate(const GlobalState &gs, const TypeConstraint &tc) c
     return Types::top();
 }
 
-TypePtr TupleType::_instantiate(const GlobalState &gs, const InlinedVector<SymbolRef, 4> &params,
-                                const vector<TypePtr> &targs) const {
-    bool changed = false;
-    vector<TypePtr> newElems;
-    newElems.reserve(this->elems.size());
-    for (auto &a : this->elems) {
-        auto t = a._instantiate(gs, params, targs);
-        if (changed || t) {
-            changed = true;
-            if (!t) {
-                t = a;
+namespace {
+template <typename... MethodArgs>
+optional<vector<TypePtr>> mungeElems(const vector<TypePtr> &elems,
+                                     TypePtr (TypePtr::*method)(const MethodArgs &...) const,
```

Comment:
i think this is the first function pointer to a method that i've seen in sorbet's codebase. not sure how i feel about it. do you think we could just implement this twice, once for _instantiate and once for _approximate?

Reply:
I am always unreasonably happy when I find a legitimate place to use pointers to methods (which happens maybe once a year).  Do you have particular uneasynesses (sp?) that you could share?  I feel like in a codebase that already takes advantage of `if constexpr`, variable-arg templates, and sfinae metaprogramming (among other things), pointers to methods are a relatively minor thing.  But I could but convinced otherwise.

Reply:
I mostly just don't like the indirection. It breaks jump to definition, where now i have to see all the call sites to understand what this method could be doing (and in fact that's what I did when reviewing: i looked at the call sites to see that it's only ever one of `_instantiate` or `_approximate`).

I think that luckily our usage of variable arg templates is currently limited to forwarding args to `fmt` (for things like `Exception::raise`)

also idk what `sp?` means, if you could expand that for me

Reply:
What about accepting a `std::function<TypePtr()>` instead, and putting the method call in the body of a lambda? Given that it wouldn't be a template anymore, it might be worth checking on godbolt to see how it inlines.

Reply:
Oh, I suppose it would need to be `std::function<TypePtr(TypePtr &)>`, but still :)

Reply:
Indirecting through `std::function` looks pretty bad: https://gcc.godbolt.org/z/E31c8T  It inlines, at least, but produces a bunch of other garbage that the compiler can't seem to get rid of.

Making jump-to-definition not work seems reasonable.  I can make that change.

> also idk what `sp?` means, if you could expand that for me

Oh, sorry, "spelling?".